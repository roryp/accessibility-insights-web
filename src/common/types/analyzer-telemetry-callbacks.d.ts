// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AxeAnalyzerResult } from '../../injected/analyzers/analyzer';
import {
    RuleAnalyzerScanTelemetryData,
    IssuesAnalyzerScanTelemetryData,
    NeedsReviewAnalyzerScanTelemetryData,
} from '../extension-telemetry-events';

export type IAnalyzerTelemetryCallback = ForRuleAnalyzerScanCallback;

export type ForRuleAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testName: string,
    requirementName?: string,
) => RuleAnalyzerScanTelemetryData;

export type ForIssuesAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testName: string,
) => IssuesAnalyzerScanTelemetryData;

export type ForNeedsReviewAnalyzerScanCallback = (
    analyzerResult: AxeAnalyzerResult,
    scanDuration: number,
    elementsScanned: number,
    testName: string,
) => NeedsReviewAnalyzerScanTelemetryData;
