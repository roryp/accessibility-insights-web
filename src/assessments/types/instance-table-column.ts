// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AssessmentInstanceRowData } from 'DetailsView/components/assessment-instance-table';

export interface InstanceTableColumn {
    key: string;
    name: string;
    onRender: (item: AssessmentInstanceRowData) => JSX.Element;
}
