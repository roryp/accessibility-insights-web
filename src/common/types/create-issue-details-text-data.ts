// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { GuidanceLink } from '../../scanner/rule-to-links-mappings';

export interface CreateIssueDetailsTextData {
    rule: {
        id: string;
        description: string;
        url: string;
        guidance: GuidanceLink[];
    };
    targetApp: {
        name: string;
        url?: string;
    };
    element: {
        identifier: string;
        conciseName: string;
    };
    howToFixSummary: string;
    snippet?: string;
}
