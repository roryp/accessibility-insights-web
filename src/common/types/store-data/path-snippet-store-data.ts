// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

export interface PathSnippetStoreData {
    path: string | null;
    snippet: string | null;
}
