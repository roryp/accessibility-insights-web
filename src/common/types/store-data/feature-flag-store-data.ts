// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { DictionaryStringTo } from 'types/common-types';

export type FeatureFlagStoreData = DictionaryStringTo<boolean>;
