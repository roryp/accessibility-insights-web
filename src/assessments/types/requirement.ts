// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { UniquelyIdentifiableInstances } from 'background/instance-identifier-generator';
import { ManualTestStatus } from 'common/types/manual-test-status';
import {
    AssessmentNavState,
    GeneratedAssessmentInstance,
    InstanceIdToInstanceDataMap,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { DetailsViewActionMessageCreator } from 'DetailsView/actions/details-view-action-message-creator';
import {
    AssessmentInstanceRowData,
    AssessmentInstanceTable,
} from 'DetailsView/components/assessment-instance-table';
import { RequirementLink } from 'DetailsView/components/requirement-link';
import { Analyzer } from 'injected/analyzers/analyzer';
import { AnalyzerProvider } from 'injected/analyzers/analyzer-provider';
import { DecoratedAxeNodeResult } from 'injected/scanner-utils';
import {
    PropertyBags,
    VisualizationInstanceProcessorCallback,
} from 'injected/visualization-instance-processor';
import { Drawer } from 'injected/visualization/drawer';
import { DrawerProvider } from 'injected/visualization/drawer-provider';
import { IColumn } from 'office-ui-fabric-react';
import { DictionaryStringTo } from 'types/common-types';
import { ContentPageComponent, HyperlinkDefinition } from 'views/content/content-page';
import { IGetMessageGenerator } from '../assessment-default-message-generator';
import { InstanceTableColumn } from './instance-table-column';
import { ReportInstanceFields } from './report-instance-field';

export interface Requirement {
    key: string;
    description: JSX.Element;
    order?: number;
    name: string;
    renderStaticContent?: () => JSX.Element;
    howToTest: JSX.Element;
    addFailureInstruction?: string;
    infoAndExamples?: ContentPageComponent;
    isManual: boolean;
    // This is for semi-manual cases where we can't present a list of instances like an assisted
    // test would, but can infer a PASS or FAIL state. If not specified, acts like () => UNKNOWN.
    getInitialManualTestStatus?: (instances: InstanceIdToInstanceDataMap) => ManualTestStatus;
    guidanceLinks: HyperlinkDefinition[];
    columnsConfig?: InstanceTableColumn[];
    getAnalyzer?: (provider: AnalyzerProvider) => Analyzer;
    getVisualHelperToggle?: (props: VisualHelperToggleConfig) => JSX.Element;
    // Any results this returns false for will be omitted from visual helper displays, but still
    // present for the purposes of instance lists or getInitialManualTestStatus. By default, all
    // results support visualization.
    isVisualizationSupportedForResult?: (result: DecoratedAxeNodeResult) => boolean;
    visualizationInstanceProcessor?: VisualizationInstanceProcessorCallback<
        PropertyBags,
        PropertyBags
    >;
    getDrawer?: (provider: DrawerProvider, featureFlagStoreData?: FeatureFlagStoreData) => Drawer;
    getNotificationMessage?: (selectorMap: DictionaryStringTo<any>) => string;
    doNotScanByDefault?: boolean;
    switchToTargetTabOnScan?: boolean;
    generateInstanceIdentifier?: (instance: UniquelyIdentifiableInstances) => string;
    reportInstanceFields?: ReportInstanceFields;
    renderReportDescription?: () => JSX.Element;
    getInstanceStatus?: (result: DecoratedAxeNodeResult) => ManualTestStatus;
    getInstanceStatusColumns?: () => Readonly<IColumn>[];
    getDefaultMessage?: IGetMessageGenerator;
    renderInstanceTableHeader?: (
        table: AssessmentInstanceTable,
        items: AssessmentInstanceRowData[],
    ) => JSX.Element;
    renderRequirementDescription?: (requirementLink: RequirementLink) => JSX.Element;
}

export type VisualHelperToggleConfigDeps = {
    detailsViewActionMessageCreator: DetailsViewActionMessageCreator;
};

export interface VisualHelperToggleConfig {
    deps: VisualHelperToggleConfigDeps;
    assessmentNavState: AssessmentNavState;
    instancesMap: DictionaryStringTo<GeneratedAssessmentInstance>;
    isStepEnabled: boolean;
    isStepScanned: boolean;
}
