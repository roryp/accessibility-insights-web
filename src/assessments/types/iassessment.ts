// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { InitialDataCreator } from 'background/create-initial-assessment-test-data';
import { RequirementOrdering } from 'common/assessment/requirement';
import { AssessmentVisualizationConfiguration } from 'common/configs/assessment-visualization-configuration';
import { AnyExtension } from 'common/extensibility/extension-point';
import { VisualizationType } from 'common/types/visualization-type';
import { ContentPageComponent } from 'views/content/content-page';
import { Requirement } from './requirement';

interface BaseAssessment {
    key: string;
    visualizationType: VisualizationType;
    title: string;
    gettingStarted: JSX.Element;
    guidance?: ContentPageComponent;
    requirements: Requirement[];
    featureFlag?: { required?: string[] };
    requirementOrder?: RequirementOrdering;
    extensions?: AnyExtension[];
    initialDataCreator?: InitialDataCreator;
}

export interface ManualAssessment extends BaseAssessment {}

export interface AssistedAssessment extends BaseAssessment {
    storeDataKey: string;
    visualizationConfiguration?: Partial<AssessmentVisualizationConfiguration>;
}

export interface Assessment extends BaseAssessment {
    getVisualizationConfiguration: () => AssessmentVisualizationConfiguration;
    requirementOrder: RequirementOrdering;
}
