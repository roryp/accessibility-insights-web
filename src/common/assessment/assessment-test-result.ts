// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { AssessmentsProvider } from 'assessments/types/assessments-provider';
import { Assessment } from 'assessments/types/iassessment';
import {
    outcomeStatsFromManualTestStatus,
    RequirementOutcomeStats,
} from 'reports/components/requirement-outcome-type';
import { ManualTestStatusData } from '../types/manual-test-status';
import { AssessmentData } from '../types/store-data/assessment-result-data';
import { VisualizationType } from '../types/visualization-type';
import { getRequirementsResults, RequirementResult } from './requirement';

export type AssessmentTestDefinition = Assessment;

export type AssessmentTestProviderDeps = {
    outcomeStatsFromManualTestStatus: (
        testStepStatus: ManualTestStatusData,
    ) => RequirementOutcomeStats;
    getRequirementsResults: (
        provider: AssessmentsProvider,
        visualizationType: VisualizationType,
        stepStatus: ManualTestStatusData,
    ) => RequirementResult[];
};

const depDefaults = {
    outcomeStatsFromManualTestStatus,
    getRequirementsResults,
};

export class AssessmentTestResult {
    constructor(
        private readonly assessmentProvider: AssessmentsProvider,
        public readonly visualizationType: VisualizationType,
        public readonly data: AssessmentData,
        private readonly deps: AssessmentTestProviderDeps = depDefaults,
    ) {}

    public getRequirementResult(requirementKey: string): RequirementResult {
        return this.getRequirementResults().find(r => r.definition.key === requirementKey);
    }

    public getRequirementResults(): RequirementResult[] {
        return this.deps.getRequirementsResults(
            this.assessmentProvider,
            this.visualizationType,
            this.data.testStepStatus,
        );
    }

    public getOutcomeStats(): RequirementOutcomeStats {
        return this.deps.outcomeStatsFromManualTestStatus(this.data.testStepStatus);
    }

    public get definition(): AssessmentTestDefinition {
        return this.assessmentProvider.forType(this.visualizationType);
    }
}
