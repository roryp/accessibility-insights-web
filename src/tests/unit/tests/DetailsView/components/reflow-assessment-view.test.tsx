// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import { Requirement } from 'assessments/types/requirement';
import { AssessmentTestResult } from 'common/assessment/assessment-test-result';
import { RequirementData, RequirementResult } from 'common/assessment/requirement';
import {
    AssessmentData,
    gettingStartedSubview,
} from 'common/types/store-data/assessment-result-data';
import { FeatureFlagStoreData } from 'common/types/store-data/feature-flag-store-data';
import { PathSnippetStoreData } from 'common/types/store-data/path-snippet-store-data';
import { AssessmentInstanceTableHandler } from 'DetailsView/handlers/assessment-instance-table-handler';
import { shallow } from 'enzyme';
import * as React from 'react';
import {
    ReflowAssessmentView,
    ReflowAssessmentViewDeps,
    ReflowAssessmentViewProps,
} from '../../../../../DetailsView/components/reflow-assessment-view';

describe('AssessmentViewTest', () => {
    let assessmentDataStub = {} as AssessmentData;

    let requirementStub: Requirement;
    let requirementStubTwo: Requirement;
    let requirementResultStub: RequirementResult;
    let requirementResultStubTwo: RequirementResult;
    let assessmentTestResultStub: AssessmentTestResult;

    let assessmentInstanceTableHandlerStub: AssessmentInstanceTableHandler;

    let featureFlagStoreDataStub: FeatureFlagStoreData;
    let pathSnippetStoreDataStub: PathSnippetStoreData;

    let reflowProps: ReflowAssessmentViewProps;

    beforeEach(() => {
        assessmentDataStub = {} as AssessmentData;

        requirementStub = {
            name: 'test-requirement-name',
            description: <div>test-description</div>,
            howToTest: <p>how-to-test-stub</p>,
            order: 1,
        } as Requirement;
        requirementStubTwo = {
            name: 'test-requirement-name-2',
            description: <div>test-description</div>,
            howToTest: <p>how-to-test-stub</p>,
            order: 2,
        } as Requirement;
        requirementResultStub = {
            definition: requirementStub,
            data: { isStepScanned: true } as RequirementData,
        };
        requirementResultStubTwo = {
            definition: requirementStubTwo,
            data: { isStepScanned: true } as RequirementData,
        };
        assessmentTestResultStub = {
            definition: {
                gettingStarted: <h1>Hello</h1>,
                title: 'some title',
                guidance: { pageTitle: 'some page title' },
            },
            getRequirementResult: _ => requirementResultStub,
            getRequirementResults: () => [requirementResultStub, requirementResultStubTwo],
            visualizationType: -1,
        } as AssessmentTestResult;

        assessmentInstanceTableHandlerStub = {
            changeRequirementStatus: null,
        } as AssessmentInstanceTableHandler;

        featureFlagStoreDataStub = {
            'some feature flag': true,
        };
        pathSnippetStoreDataStub = {
            path: null,
        } as PathSnippetStoreData;

        reflowProps = {
            deps: {} as ReflowAssessmentViewDeps,
            prevTarget: { id: 4 },
            currentTarget: { id: 5 },
            scanningInProgress: true,
            selectedRequirementIsEnabled: true,
            assessmentNavState: {
                selectedTestType: -1,
            },
            assessmentData: assessmentDataStub,
            assessmentTestResult: assessmentTestResultStub,
            assessmentInstanceTableHandler: assessmentInstanceTableHandlerStub,
            featureFlagStoreData: featureFlagStoreDataStub,
            pathSnippetStoreData: pathSnippetStoreDataStub,
        } as ReflowAssessmentViewProps;
    });

    test('render for requirement', () => {
        reflowProps.assessmentNavState.selectedTestSubview = 'requirement';
        const rendered = shallow(<ReflowAssessmentView {...reflowProps} />);
        expect(rendered.getElement()).toMatchSnapshot();
    });

    test('render for getting started', () => {
        reflowProps.assessmentNavState.selectedTestSubview = gettingStartedSubview;
        const rendered = shallow(<ReflowAssessmentView {...reflowProps} />);
        expect(rendered.getElement()).toMatchSnapshot();
    });

    test('render for requirement where nextRequirement is null', () => {
        reflowProps.assessmentNavState.selectedTestSubview = 'requirement';
        reflowProps.assessmentTestResult.getRequirementResult = _ => requirementResultStubTwo;
        const rendered = shallow(<ReflowAssessmentView {...reflowProps} />);
        expect(rendered.getElement()).toMatchSnapshot();
    });
});
