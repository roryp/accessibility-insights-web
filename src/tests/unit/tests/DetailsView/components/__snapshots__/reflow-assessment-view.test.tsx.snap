// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssessmentViewTest render for getting started 1`] = `
<React.Fragment>
  <TargetChangeDialog
    deps={Object {}}
    newTab={
      Object {
        "id": 5,
      }
    }
    prevTab={
      Object {
        "id": 4,
      }
    }
  />
  <GettingStartedView
    currentTest={-1}
    deps={Object {}}
    gettingStartedContent={
      <h1>
        Hello
      </h1>
    }
    guidance={
      Object {
        "pageTitle": "some page title",
      }
    }
    nextRequirement={
      Object {
        "description": <div>
          test-description
        </div>,
        "howToTest": <p>
          how-to-test-stub
        </p>,
        "name": "test-requirement-name",
        "order": 1,
      }
    }
    title="some title"
  />
</React.Fragment>
`;

exports[`AssessmentViewTest render for requirement 1`] = `
<React.Fragment>
  <TargetChangeDialog
    deps={Object {}}
    newTab={
      Object {
        "id": 5,
      }
    }
    prevTab={
      Object {
        "id": 4,
      }
    }
  />
  <RequirementView
    assessmentData={Object {}}
    assessmentInstanceTableHandler={
      Object {
        "changeRequirementStatus": null,
      }
    }
    assessmentNavState={
      Object {
        "selectedTestSubview": "requirement",
        "selectedTestType": -1,
      }
    }
    currentTarget={
      Object {
        "id": 5,
      }
    }
    deps={Object {}}
    featureFlagStoreData={
      Object {
        "some feature flag": true,
      }
    }
    isRequirementEnabled={true}
    isRequirementScanned={true}
    nextRequirement={
      Object {
        "description": <div>
          test-description
        </div>,
        "howToTest": <p>
          how-to-test-stub
        </p>,
        "name": "test-requirement-name-2",
        "order": 2,
      }
    }
    pathSnippetStoreData={
      Object {
        "path": null,
      }
    }
    prevTarget={
      Object {
        "id": 4,
      }
    }
    requirement={
      Object {
        "description": <div>
          test-description
        </div>,
        "howToTest": <p>
          how-to-test-stub
        </p>,
        "name": "test-requirement-name",
        "order": 1,
      }
    }
    scanningInProgress={true}
  />
</React.Fragment>
`;

exports[`AssessmentViewTest render for requirement where nextRequirement is null 1`] = `
<React.Fragment>
  <TargetChangeDialog
    deps={Object {}}
    newTab={
      Object {
        "id": 5,
      }
    }
    prevTab={
      Object {
        "id": 4,
      }
    }
  />
  <RequirementView
    assessmentData={Object {}}
    assessmentInstanceTableHandler={
      Object {
        "changeRequirementStatus": null,
      }
    }
    assessmentNavState={
      Object {
        "selectedTestSubview": "requirement",
        "selectedTestType": -1,
      }
    }
    currentTarget={
      Object {
        "id": 5,
      }
    }
    deps={Object {}}
    featureFlagStoreData={
      Object {
        "some feature flag": true,
      }
    }
    isRequirementEnabled={true}
    isRequirementScanned={true}
    pathSnippetStoreData={
      Object {
        "path": null,
      }
    }
    prevTarget={
      Object {
        "id": 4,
      }
    }
    requirement={
      Object {
        "description": <div>
          test-description
        </div>,
        "howToTest": <p>
          how-to-test-stub
        </p>,
        "name": "test-requirement-name-2",
        "order": 2,
      }
    }
    scanningInProgress={true}
  />
</React.Fragment>
`;
